token.c

```diff
@@ -0,0 +1,153 @@
+/****
+ *
+ * Compile with
+ * gcc -o token -lcrypto -lcurl -Wno-deprecated-declarations token.c
+ *
+ ****/
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/time.h>
+#include <curl/curl.h>
+#include <openssl/bio.h>
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+
+struct string {
+    char *ptr;
+    size_t len;
+};
+
+static const unsigned char* encoded_key = "RmFzdGx5IFRva2VuIFRlc3Q=";
+static const unsigned long  interval    = 60;
+
+unsigned char* decode64(const unsigned char *input, int length);
+unsigned char* encode64(const unsigned char *input, int length);
+unsigned char* pack64(long value);
+void init_string(struct string *s);
+void free_string(struct string *s);
+int fetch_url(const char* url, struct string *s);
+
+int main(int argc, char** argv) {
+    struct timeval tv;
+    gettimeofday(&tv, NULL);
+    
+    const unsigned char* key    = decode64(encoded_key, strlen(encoded_key));
+    const unsigned char* number = pack64((long) tv.tv_sec / interval);
+    
+    OpenSSL_add_all_digests();
+    unsigned int md_len;
+    unsigned char* digest = HMAC(EVP_sha256(), key, strlen(key), number, 8, NULL, &md_len);
+    unsigned char* token  = encode64(digest, md_len);
+    
+    struct string response;
+    init_string(&response);
+    fetch_url("http://token.fastly.com/token", &response);
+    
+    struct string validation;
+    init_string(&validation);
+    char* url = calloc(1024, sizeof(char));
+    sprintf(url, "http://token.fastly.com?%s", token);
```

Comment:
prefer `snprintf` over `sprintf`, especially since we know the buffer size.  Or, if you want to be real neat, use `asprintf` and avoid the allocation above.